## In these exercises, you will complete the first step toward the above goal: acquire and prepare the necessary Zillow data from the zillow database in the Codeup database server.

1. Acquire bedroomcnt, bathroomcnt, calculatedfinishedsquarefeet, taxvaluedollarcnt, yearbuilt, taxamount, and fips from the zillow database for all 'Single Family Residential' properties.
2. Using your acquired Zillow data, walk through the summarization and cleaning steps in your wrangle.ipynb file like we did above. You may handle the missing values however you feel is appropriate and meaningful; remember to document your process and decisions using markdown and code commenting where helpful.
3. Store all of the necessary functions to automate your process from acquiring the data to returning a cleaned dataframe with no missing values in your wrangle.py file. Name your final function wrangle_zillow.

# Acquire

In [1]:
import pandas as pd 
import env
from sklearn.model_selection import train_test_split
def get_db_url(database):
    return f'mysql+pymysql://{env.username}:{env.password}@{env.host}/{database}'

In [2]:
url=get_db_url('zillow')
    
SQL_query= """select id, parcelid,bathroomcnt, bedroomcnt,calculatedfinishedsquarefeet
        ,taxvaluedollarcnt,yearbuilt,taxamount, fips
         from properties_2017
         where propertylandusetypeid=261"""
    
filename=('zillow.csv')

directory='/Users/chellyannmoreno/codeup-data-science/regression-exercises/'

In [4]:
def get_data():
    if os.path.exists(directory+filename):
        df=pd.read_csv(filename)
        return df
    else:
        df=pd.read_sql(SQL_query,url)
        df.to_csv(filename,index=False)
        return df

In [12]:
df=get_data()

In [13]:
df.head()

id  parcelid  bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet  \
0   1  10759547          0.0         0.0                           NaN   
1  15  11193347          0.0         0.0                           NaN   
2  16  11215747          0.0         0.0                           NaN   
3  17  11229347          0.0         0.0                           NaN   
4  20  11324547          2.0         4.0                        3633.0   

   taxvaluedollarcnt  yearbuilt  taxamount    fips  
0            27516.0        NaN        NaN  6037.0  
1               10.0        NaN        NaN  6037.0  
2               10.0        NaN        NaN  6037.0  
3             2108.0        NaN     174.21  6037.0  
4           296425.0     2005.0    6941.39  6037.0

In [14]:
df.shape

(2152863, 9)

# Prepare

In [51]:
df.shape

(2028095, 9)

In [17]:
oldh=df[df.yearbuilt<1900]
oldh.shape

(2561, 9)

In [19]:
oldh.yearbuilt.value_counts()

1895.0    569
1890.0    539
1898.0    256
1885.0    172
1899.0    139
1897.0    135
1896.0    128
1894.0    103
1893.0     91
1888.0     88
1887.0     78
1892.0     65
1880.0     39
1886.0     33
1891.0     31
1889.0     30
1884.0     14
1875.0     10
1883.0      8
1882.0      6
1881.0      4
1870.0      3
1801.0      3
1812.0      2
1865.0      2
1876.0      2
1878.0      2
1862.0      1
1807.0      1
1877.0      1
1879.0      1
1823.0      1
1855.0      1
1833.0      1
1840.0      1
1874.0      1
Name: yearbuilt, dtype: int64

In [71]:
len(oldh)/len(df)

0.0011991548719364725

In [28]:
df.describe().T

count          mean           std          min           25%  \
id         2033172.0  1.490903e+06  8.608481e+05        20.00  7.455758e+05   
parcelid   2033172.0  1.304209e+07  4.344449e+06  10711725.00  1.167093e+07   
bathrooms  2033172.0  2.141026e+00  8.351854e-01         0.00  2.000000e+00   
bedrooms   2033172.0  3.257700e+00  8.967887e-01         0.00  3.000000e+00   
area       2033172.0  1.751421e+03  7.258099e+02         1.00  1.242000e+03   
taxvalue   2033172.0  3.632401e+05  2.432811e+05        22.00  1.816560e+05   
yearbuilt  2033172.0  1.960267e+03  2.154476e+01      1801.00  1.949000e+03   
taxamount  2033172.0  4.506123e+03  2.842968e+03         6.34  2.456707e+03   

                   50%           75%           max  
id          1490529.00  2.235772e+06  2.982282e+06  
parcelid   12610664.50  1.404881e+07  1.681836e+08  
bathrooms         2.00  3.000000e+00  2.000000e+01  
bedrooms          3.00  4.000000e+00  1.600000e+01  
area           1587.00  2.106000e+03  2.422000e+04  
taxvalue     312745.00  4.917210e+05  1.202999e+06  
yearbuilt      1958.00  1.974000e+03  2.016000e+03  
taxamount      3937.89  5.938602e+03  1.003793e+05

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2028095 entries, 4 to 2152862
Data columns (total 9 columns):
 #   Column     Dtype  
---  ------     -----  
 0   id         int64  
 1   parcelid   int64  
 2   bathrooms  float64
 3   bedrooms   int64  
 4   area       int64  
 5   taxvalue   int64  
 6   yearbuilt  int64  
 7   taxamount  float64
 8   county     object 
dtypes: float64(2), int64(6), object(1)
memory usage: 154.7+ MB


In [24]:
df.bathrooms.value_counts()

2.00     931071
1.00     408968
3.00     400667
2.50     140045
4.00      62921
1.50      31005
3.50      22502
5.00      14210
4.50      10614
6.00       2830
5.50       1229
7.00        616
6.50        150
8.00        134
7.50         63
9.00         28
10.00        15
0.50         12
11.00         5
1.75          3
9.50          3
14.00         2
12.00         1
8.50          1
13.00         1
Name: bathrooms, dtype: int64

In [20]:
df = df.rename(columns={
        'bedroomcnt': 'bedrooms',
        'bathroomcnt': 'bathrooms',
        'calculatedfinishedsquarefeet': 'area',
        'taxvaluedollarcnt': 'taxvalue',
        'fips': 'county'
    })

    # Filter out rows with large area and filter out places with zero bathrooms and baths, and with more than 15.
df = df[df.area < 25_000]
df = df[df.yearbuilt > 1890]
df = df[(df.bathrooms > 0) & (df.bathrooms < 15) & (df.bedrooms > 0) & (df.bedrooms < 15)]


    # Drop rows with missing values
df = df.dropna()

    # Filter out rows with high tax value
taxvalue_threshold = df.taxvalue.quantile(.95)
df = df[df.taxvalue < taxvalue_threshold].copy()

    # Convert data types
df[['bedrooms', 'area', 'taxvalue', 'yearbuilt']] = df[['bedrooms', 'area', 'taxvalue', 'yearbuilt']].astype(int)

    # Map county codes to names
county_map = {6037: 'LA', 6059: 'Orange', 6111: 'Ventura'}
df.county = df.county.map(county_map)

In [21]:
df.head()

id  parcelid  bathrooms  bedrooms  area  taxvalue  yearbuilt  taxamount  \
4    20  11324547        2.0         4  3633    296425       2005    6941.39   
6    31  11544747        4.0         3  1620    847770       2011   10244.94   
7    33  11585547        2.0         3  2077    646760       1926    7924.68   
18  110  12716947        1.0         3  1244    169471       1950    2532.88   
19  111  12757147        2.0         3  1300    233266       1950    3110.99   

   county  
4      LA  
6      LA  
7      LA  
18     LA  
19     LA

In [22]:
df.shape

(2027096, 9)

In [23]:
2027096/2152863

0.9415815126183134

# Recap:
1. created my acquire data function
2. drop all nulls
3. remove houses with areas of more than 25_000
4. remove houses with no beds and/or baths
5. remove houses that are older than 1880
6. rename columns
7. county column turned into object
8. verify if they were any halfbaths
9. after all changes were made I had kept around 94% of my data.

# Split data

In [ ]:
# Split into train_validate and test sets
train_validate, test = train_test_split(df, test_size=.2, random_state=123)

# Split into train and validate sets
train, validate = train_test_split(train_validate, test_size=.25, random_state=123)

